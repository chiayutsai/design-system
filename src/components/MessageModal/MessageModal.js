/* eslint css-modules/no-unused-class: [2, { markAsUsed: ['error',
'info', 'message', 'notice', 'confirm', 'cancel'] }] */
import classnames from 'classnames'
import useStyles from 'isomorphic-style-loader/useStyles'
import PropTypes from 'prop-types'
import { useCallback } from 'react'
import { useTranslation } from 'react-i18next'
import { useDispatch } from 'react-redux'
import { dismissSystemMessage } from 'actions/message'
import { BUTTON_CONTENT_MAP } from 'constants/buttonType'
import ModalWrapper from '../ModalWrapper'
import { ReactComponent as IconNoticeSvg } from './assets/icon-notice.svg'
import styles from './MessageModal.scss'

const MessageModal = ({
  id,
  type,
  code,
  buttons,
  messageText,
  messageContent,
  shouldCloseOnOverlayClick,
}) => {
  useStyles(styles)
  const { t } = useTranslation()

  const dispatch = useDispatch()

  const onClose = useCallback(() => {
    dispatch(dismissSystemMessage({ id }))
  }, [dispatch, id])

  const handleButtonClick = useCallback(
    ({ currentTarget: el }) => {
      const index = [...el.parentElement.children].indexOf(el)
      const { onClick } = buttons[index]
      if (onClick) {
        onClick(onClose, index)
      } else {
        onClose()
      }
    },
    [buttons, onClose],
  )

  return (
    <ModalWrapper
      className={styles.modal}
      overlayClassName={styles.overlay}
      shouldCloseOnOverlayClick={shouldCloseOnOverlayClick}
      onClose={onClose}>
      <div className={styles.container}>
        <IconNoticeSvg className={classnames(styles.icon, styles[type])} />
        <div className={styles['message-content']}>
          {messageContent ? t(messageContent) : messageText}
          {!!code && `[${code}]`}
        </div>
        <div className={styles['button-group']}>
          {buttons?.map(({ label, content }) => (
            <div
              key={`button-${label}`}
              className={classnames(styles.button, styles[label.toLowerCase()])}
              role="presentation"
              onClick={handleButtonClick}>
              {t(content || BUTTON_CONTENT_MAP[label])}
            </div>
          ))}
        </div>
      </div>
    </ModalWrapper>
  )
}

MessageModal.propTypes = {
  id: PropTypes.number.isRequired,
  buttons: PropTypes.arrayOf(PropTypes.object),
  type: PropTypes.string.isRequired,
  code: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  messageContent: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
  messageText: PropTypes.string,
  shouldCloseOnOverlayClick: PropTypes.bool,
}
MessageModal.defaultProps = {
  code: 0,
  buttons: [],
  shouldCloseOnOverlayClick: false,
  messageContent: '',
  messageText: '',
}

export default MessageModal
