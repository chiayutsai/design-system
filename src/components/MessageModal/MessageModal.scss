.modal {
  align-items: flex-start;

  .overlay {
    background-color: rgba(0, 0, 0, 0.5);
  }
}

.container {
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 294px;
  min-height: 300px;
  padding: 20px 27px;
  border-radius: 4%;
  color: var(--wrong-pop-2);
  background-color: var(--wrong-pop-1);
  box-shadow: 0 0 20px 0 rgba(0, 0, 0, 0.3);

  .icon {
    width: 70px;
    height: 70px;
    margin: 15px 0;

    &.error,
    &.info,
    &.message,
    &.notice {
      g {
        g {
          stroke: var(--wrong-pop-2);
        }

        text {
          fill: var(--wrong-pop-2);
        }
      }
    }
  }

  .message-content {
    display: flex;
    align-items: center;
    width: 100%;
    min-height: 100px;
    padding: 0 6px;
    font-size: 16px;
  }

  .button-group {
    display: flex;
    width: 100%;
    height: 40px;
    margin-top: 20px;

    .button {
      display: flex;
      justify-content: center;
      align-items: center;
      width: 100%;
      height: 100%;
      border-radius: 4px;

      &:not(:first-child) {
        margin-left: 10px;
      }


      &.confirm {
        color: var(--wrong-pop-1);
        background-color: var(--wrong-pop-3);
      }

      &.cancel {
        color: var(--wrong-pop-3);
        background-color: var(--wrong-pop-1);
      }

      &:hover {
        background-color: var(--footer-button-7);
        color: var(--footer-button-9);
        transition: 0.3s;
      }

      &:active {
        background-color: var(--footer-button-7);
        color: var(--footer-button-10);
        transform: scale(0.96);
        font-weight: bolder;
      }
    }
  }
}
