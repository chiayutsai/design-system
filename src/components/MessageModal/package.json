{
  "name": "MessageModal",
  "version": "0.0.0",
  "private": true,
  "main": "MessageModal.js"
}
