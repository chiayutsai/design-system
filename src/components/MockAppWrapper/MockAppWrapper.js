import history from 'history'
import PropTypes from 'prop-types'
import App from 'components/App'
import configureStore from 'store/configureStore'
import { injectReducer } from 'store/makeRootReducer'

const initialState = {}

export const store = configureStore(initialState, { history })

const MockAppWrapper = ({ children }) => (
  <App
    context={{
      pathname: '',
      store,
      injectReducer,
    }}
    insertCss={() => {}}>
    {children}
  </App>
)

MockAppWrapper.propTypes = {
  children: PropTypes.node.isRequired,
}

export default MockAppWrapper
