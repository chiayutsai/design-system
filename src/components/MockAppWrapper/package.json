{
  "name": "MockAppWrapper",
  "version": "0.0.0",
  "private": true,
  "main": "./MockAppWrapper.js"
}
