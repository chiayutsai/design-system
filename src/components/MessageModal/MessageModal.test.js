import renderer from 'react-test-renderer'
import MockAppWrapper, { store } from 'components/MockAppWrapper'
import setupLocale from 'test/utils/setupLocale'
import MessageModal from './MessageModal'

describe('MessageModal', () => {
  // setup i18next if component needed
  setupLocale(store)

  test('render correctly', () => {
    const wrapper = renderer
      .create(
        <MockAppWrapper>
          <MessageModal id={1} type="message" />
        </MockAppWrapper>,
      )
      .toJSON()

    expect(wrapper).toMatchSnapshot()
  })
})
