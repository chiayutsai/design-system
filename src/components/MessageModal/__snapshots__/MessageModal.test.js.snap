// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageModal render correctly 1`] = `
<div
  className="root modal overlay"
  onClick={[Function]}
  role="presentation"
>
  <div
    className="container"
  >
    <div
      className="icon message"
    />
    <div
      className="message-content"
    >
      
    </div>
    <div
      className="button-group"
    />
  </div>
</div>
`;
